### PR TITLE
Annotate `WebAppConfig` to fix Gradle Deprecation Warnings

### PR DESCRIPTION
**Pull request by f4lco**
I did some reading & analysis and this fixes a couple of warnings from #133. I'm very unsure about this solution, please see for yourself and decide if we need to rethink. I dislike:

- the amount of boilerplate added: if it weren't for the `@Delegate`, we could just annotate the Groovy property as in:  
  ```groovy
  class WebAppConfig {
    @Input
    @Optional
    def contextPath
    // ...
  }
  ```
- making "gretty-core" dependent on Gradle's API, polluting those data transfer objects with Gradle-specific input and output logic.

@boris-petrov I made this a draft PR - please tell me what you think and then we figure out how to proceed.

### CODE REVIEW

**Comment by boris-petrov** [comment]
@f4lco - thank you, yet again, for taking the time to dig into this. I keep postponing it but yes, Gradle 7 is coming and something must be done, so thanks again!

I am (almost) completely unfamiliar with Gradle's APIs. Could you briefly explain to me a few things?

`org.gradle.warning.mode=all` - this I understand, I have the same in my own project.

`implementation gradleApi()` - I'm not sure what this means, why is it needed, why it worked without it and which warnings does it fix? Could you elaborate?

Also I'm not sure what `@Delegate` do you mean because of which you cannot put the `Input` and `Optional` annotations directly on the field? Otherwise I'm totally fine with the boilerplate if there is no other way.

I actually tried a few times to tackle these deprecation warnings and every time I couldn't even start. Most of the fields mentioned [here](https://github.com/gretty-gradle-plugin/gretty/issues/133#issue-530718082) I couldn't even find when grepping and the one [here](https://github.com/gretty-gradle-plugin/gretty/issues/133#issuecomment-560255052) is going to be the really nasty one. It reminds me of the horrible [fixes](https://github.com/gretty-gradle-plugin/gretty/commit/39f3fa8387ae330822fa653839bf33a7aa40b605) we had to [do](https://github.com/gretty-gradle-plugin/gretty/commit/42678bbebf5c6d0b581bad86fbf6672bde83d170) for Gradle 6 and I'm really scared of how difficult this will be to resolve. :smiley:

**Comment by f4lco** [comment]
Let's see if we can unpack a bit:

> implementation gradleApi() - I'm not sure what this means, why is it needed, why it worked without it and which warnings does it fix? Could you elaborate?

That dependency does not fix any warning by itself. It's just that `WebAppConfig` resides in the module "gretty-core", which did not have access to any of Gradle's classes or annotations. So adding `gradleApi()` to the list of dependencies unlocks access to `org.gradle.api.tasks.Input` in the first place.

> Also I'm not sure what `@Delegate` do you mean because of which you cannot put the Input and Optional annotations directly on the field? 

It's the following `@Delegate`:

```groovy
class AppStartTask extends StartBaseTask {
  @Delegate(methodAnnotations = true)
  private WebAppConfig webAppConfig = new WebAppConfig()

  // ...
}
```

Gradle analyses the task `AppStartTask` to see which properties constitute inputs or outputs of the task. In order for Gradle to find the properties and their annotations, one of the following must apply:

1. Gradle finds a Groovy annotated property directly located in that class.
2. Gradle finds an annotated getter method.

In our setup (2) is the case. After compilation, the task class looks like the following:

```groovy
class AppStartTask extends StartBaseTask {
  @Delegate(methodAnnotations = true)
  private WebAppConfig webAppConfig = new WebAppConfig()

  @Generated
  @Input
  @Optional
  public Set<String> getBeforeClassPath() {
    return webAppConfig.beforeClassPath
  }

  @Generated
  public void setBeforeClassPath(Set<String> param0) {
    webAppConfig.beforeClassPath = param0
  }
  
  // repeat for all properties
  // ...
}
```

---

Parts of the explanation are now a bit redundant. I figured, if we are forced to write annotated getters within this setup, we might as well put those straight into the task. This puts task-related annotations such as input and output right in the task class where they belong. We get rid of the additional `gradleApi()` dependency and even `WebAppConfig` stays as-is. I think that is the way to go, see the last commit, "Refactor: ...".

**Comment by f4lco** [comment]
I annotated the remaining properties. This patch fixes 23 deprecation warnings in comparison to master, all originating in `WebAppConfig`. 63 to go 😬

**Comment by wolfs** [comment]
@f4lco I also starting working yesterday on fixing those warnings. I was looking at `ServerConfig` using the same approach as you do. I'll open a PR for `ServerConfig` soon.

Note that I suggest adding `gradleApi()` as `compileOnly`, given that it is not needed at runtime.

**Comment by wolfs** [comment]
Hmm... I saw you moved the annotations to the task now. I actually would prefer adding them on the classes which the tasks delegates to. Or are these classes not re-used between tasks, aka multiple tasks delegate to the same `*Config` class?

**Comment by f4lco** [comment]
Cool your also tackling this, @wolfs!

> Or are these classes not re-used between tasks, aka multiple tasks delegate to the same *Config class?

That would indeed be a concern. My analysis is as follows, these are the only tasks which embed a `serverConfig` or `webAppConfig`:

```
StartBaseTask
+- FarmStartTask
    +- serverConfig
+- AppStartTask
    +- serverConfig
    +- webAppConfig
```
So right now, it's a problem for `serverConfig`. Both `AppStartTask` and `FarmStartTask` would have to carry the annotated accessors... except if we pulled up `serverConfig` into `StartBaseTask`. The latter would be my preferred solution / suggestion - however, I have not tried it out yet!

**Comment by wolfs** [comment]
What is the downside of adding `gradleApi()` as a `compileOnly` dependency to `gretty-core`?

**Comment by wolfs** [comment]
@f4lco I created https://github.com/gretty-gradle-plugin/gretty/pull/195, WDYT?

**Comment by f4lco** [comment]
@wolfs probably there's no factual downside to that dependency in `gretty-core`, it's more of an architectural decision. Considering this,

- I think that `gretty-core` was designed without `gradleApi` in mind, and potentially intended to be reused independent of `gradleApi`, e.g. from a runner process, which also by design does not inherit the classpath bloat of Gradle.
- I assume that `compileOnly` is only part of the solution. That dependency invites coding against more classes from `gradleApi`, but then lacks runtime components. Therefore, it might crash at runtime, when loading a class that uses symbols from `gradleApi`.
- I still like that I can pop up a task class (or base class) and immediately see how the task wires inputs and outputs to properties of configuration objects (data transfer objects). To me that design is very straightforward and easier to understand.
- I would trade a bit of composability between tasks for the added clarity: not hiding task I/O in an at first glance unrelated module, no delegation-with-annotation-copying, and no "compileOnly" deps. Gretty has already a fair share of dependency conflicts, and this would add to the conundrum.

These are my 2ct. Please at least consider doing an experiment or developing an alternate solution which addresses these concerns.
TL;DR: I think the decoupling of gretty-core from Gradle's API is still valid, and adding gradleApi works against that.

**Comment by boris-petrov** [comment]
@f4lco - thanks for the explanations and thanks to @wolfs for also joining in!

After reading the full discussion, now I understand the whole situation. I agree with @f4lco with his points about `gretty-core` but going through the projects, I see there also is a `gretty-common` one which doesn't depend on Gradle stuff and all `gretty-runner*` projects depend on it. Perhaps we could use that one (or extract even a new one) for the things that @f4lco mentioned - as separation from Gradle bloat - and make `gretty-core` a "base" project for all Gradle-related things? Just an idea.

**Comment by boris-petrov** [comment]
Also, we use `implementation` instead of `compileOnly` for `gradleApi` in the [two](https://github.com/gretty-gradle-plugin/gretty/blob/master/libs/gretty/build.gradle#L15) places it is [used](https://github.com/gretty-gradle-plugin/gretty/blob/master/integrationTests/buildSrc/gretty-integrationTest/build.gradle#L6). Should we change this?

**Comment by wolfs** [comment]
> Also, we use implementation instead of compileOnly for gradleApi in the two places it is used. Should we change this?

@boris-petrov You should not change it. I used `compileOnly` in gretty-core since we only needed the annotations at compile time. We don't need `gradleApi()` at runtime there. In the above projects, you need `gradleApi()` at runtime, since you actually have some tasks which you want to run.

**Comment by wolfs** [comment]
@f4lco How about this: I introduced [an interface](https://github.com/gretty-gradle-plugin/gretty/pull/195/files#diff-617e433d6b0868429e7fce2f60eb877972e39bb69cecc31c83609dc9191de78aR16) which the tasks which delegate to `ServerConfig` can implement and which contains the annotations. This way we can re-use the annotations and we do not add the extra dependency in `gretty-core`?

**Comment by f4lco** [comment]
>  I see there also is a gretty-common one which doesn't depend on Gradle stuff and all gretty-runner* projects depend on it. Perhaps we could use that one (or even extract a new one) for the things that @f4lco mentioned - as separation from Gradle bloat - and make gretty-core a "base" project for all Gradle-related things? Just an idea.

@boris-petrov if I get you correctly, that seems like a good idea. I'll try it out on time and report back.

@wolfs nice idea, nails the concerns. I think that's also a good way to go 👍  I'll update this PR accordingly.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy [hunk 1]
```diff
@@ -25,6 +27,144 @@ class AppStartTask extends StartBaseTask {
...
+
+  @Input
+  @Optional
+  getRealm() {
+    webAppConfig.realm
+  }
+
+  @Input
+  @Optional
+  getRealmConfigFile() {
```
Comment (wolfs) [inline libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy:50]: If this is a `File`, it shouldn't be `@Input`. Gradle 7.0 will actually fail the build if you do something like this.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy [hunk 2]
```diff
@@ -25,6 +27,144 @@ class AppStartTask extends StartBaseTask {
...
+    webAppConfig.realmConfigFile
+  }
+
+  @Input
```
Comment (wolfs) [inline libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy:54]: Files shouldn't be `@Input`, but `@InputFile` with a `PathSensitivity`.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy [hunk 3]
```diff
@@ -25,6 +27,144 @@ class AppStartTask extends StartBaseTask {
...
+  @Optional
+  getContextConfigFile() {
+    webAppConfig.contextConfigFile
+  }
+
+  @Input
+  @Optional
+  getScanDirs() {
```
Comment (wolfs) [inline libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy:62]: Is this really an input? I would keep this `@Internal` or only use the path. Since using only the path is more involved (add a separate getter returning the absolute path), I would opt for ignoring this as an input for now.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy [hunk 4]
```diff
@@ -25,6 +27,144 @@ class AppStartTask extends StartBaseTask {
   @Delegate
   private WebAppConfig webAppConfig = new WebAppConfig()
 
+  @Input
+  @Optional
+  getContextPath() {
```
Comment (wolfs) [inline libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy:32]: A general comment: The return type is missing for all these methods. I think you need to add the return type, so they actually override the method from the delegate. And it would be clearer as well I think.

**Review by wolfs — COMMENTED** [review]
I left some comments. I think you still wanted to move the overrides to an interface, right? Or is this something you don't want to do for `WebAppConfig`?

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy (hunk 2 above)
Comment (f4lco) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy:54]: Thanks, will fix that.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy (hunk 1 above)
Comment (f4lco) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy:50]: Thanks for the heads-up.  This property will actually be annotated by `WebAppConfigWithInputs` as well as `ServerConfigWithInputs`. As a result, the solution is to mark it `@Internal`, to agree with the declaration in your PR.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy (hunk 3 above)
Comment (f4lco) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy:62]: OK will do.

File: libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy (hunk 4 above)
Comment (f4lco) [reply libs/gretty/src/main/groovy/org/akhikhl/gretty/AppStartTask.groovy:32]: Good call. That's reasonable, will do.

**Comment by f4lco** [comment]
@wolfs yes, I want an interface, just to mirror your approach. Please re-review, I think I addressed all of your remarks. I also added a couple of more props. I verified that, together with your PR, all properties are annotated 🎉 
Next steps: @boris-petrov will merge your PR, I clean up this one & resolve conflicts, and then merge this PR.

**Comment by boris-petrov** [comment]
@f4lco - the other branch is merged, so you can rebase, squash and then we can merge. Thanks!

**Comment by f4lco** [comment]
@boris-petrov that's it! I removed the "draft" state. I added a rename for the two interfaces - I think that name fits way better. Please merge at will.

**Review by wolfs — APPROVED** [review]
LGTM!

**Comment by boris-petrov** [comment]
OK, GitHub Actions is also happy, so thanks again @f4lco and @wolfs!
